mat64/doc.go

```diff
@@ -2,38 +2,119 @@
+// Note that some operations have a transpose as part of their definition, as in
+// *SymDense.SymOuterK.
+//
+// Matrix Factorization
+//
+// Matrix factorizations, such as the LU decomposition, typically have their own
+// specific data storage, and so are each implemented as a specific type. The
+// factorization can be computed through a call to Factorize
+//  var lu mat64.LU
+//  lu.Factorize(a)
+// The elements of the factorization can be extracted through methods on the
+// appropriate type, i.e. *TriDense.LFromLU and *TriDense.UFromLU. Alternatively,
+// they can be used directly, as in *Dense.SolveLU. Some factorizations can be
+// updated directly, without needing to update the original matrix and refactorize,
+// as in *LU.RankOne.
+//
+// BLAS and LAPACK
+//
+// BLAS and LAPACK are the standard APIs for linear algebra routines. Many
+// operations in mat64 are implemented using calls to the wrapper functions
+// in gonum/blas/blas64 and gonum/lapack/lapack64. By default, blas64 and
+// lapack64 call the native Go implementations of the routines.  Alternatively,
+// it is possible to use c-based implementations of the APIs through the respective
+// cgo packages and "Use" routines. The Go implementation of LAPACK makes calls
+// through blas64, so if a cgo BLAS implementation is registered, the lapack64
+// calls will be partially executed in Go and partially executed in c.
+//
+// Type Switching
+//
+// The Matrix abstraction enables efficiency as well as interoperability. Go's
+// type reflection capabilities are used to choose the most efficient routine
+// given the specific concrete types. For example, in
+//  c.Mul(a, b)
+// if a and b both implement RawMatrixer, that is, they can be represented as a
+// blas64.General, blas64.Gemm (general matrix mulitplication) is called, while
+// instead if b is a RawSymmetricer blas64.Symm is used (general-symmetric
+// multiplication), and if b is a *Vector, blas64.Gemv is used.
+//
+// There are many possible type combinations and special cases. No specific guarantees
+// are made about the performance of any method, and in particular, note that an
+// abstract matrix type may be copied into a concrete type of the corresponding
+// value. If there are specific special cases that are needed, please submit a
+// pull-request or file an issue.
+//
+// Invariants
+//
+// Matrix input arguments to functions are never modified. If an operation would
+// change Matrix data, that matrix will be the receiver of a function.
+//
+// For convenience, a matrix may be used as both a receiver and as an input, i.e.
+//  a.Pow(a, 6)
+//  v.SolveVec(a.T(), v)
+// though in many cases this will cause an allocation. However, mat64 cannot
+// detect arbitrary overlap, and so care should be taken when using matrix views.
+// An exception to this rule is Copy, which does not allow a.Copy(a.T()).
+//
+// All functions may be called concurrently.
```

Comment:
Probably not. In my mind, functions should always be written so they are concurrently callable, but I'm not sure what the general Go "standard wisdom" is.